### ISSUE TITLE
Is it possible to encode a 10+ bit HEIF with vips -> libheif?

### ISSUE THREAD

**Issue by AttilaTheFun** [libvips/libvips#2572]
Libheif explicitly states that it includes support for encoding 10 and 12 bit images (and I believe 16 bit as well though it's not listed):

> The encoder supports:
> ...
> 10 and 12 bit images

From: https://github.com/strukturag/libheif

Though after looking over heifsave.c I don't see a way to invoke libheif to encode an image with an explicit bit depth:
https://github.com/libvips/libvips/blob/47383b5bfc136f7870c319a1edce4848e177425b/libvips/foreign/heifsave.c

Is it currently possible to accomplish this? If so, how might I go about doing so? And if not, what might we have to change in order to support this functionality on the vips side? AFAICT PNGs are the only image type vips supports with a bit depth option when encoding. Thanks!

**Comment by jcupitt** [comment]
Hi @AttilaTheFun,

No, the libheif saver in libvips only supports 8-bit image at the moment. I've not looked at the libheif API for high bit depth images --- someone would need to read it and make a PR to libvips.

The `bitdepth` option to the libvips savers is for supporting low bitdepths (eg. 1, 2, 4 bits). For depths >8, you can just send (for example) a 16-bit image to the saver. From memory, the savers for vips, tiff, png, ppm, fits, nifti all support >8 bit pixels, perhaps there are others.

**Comment by jcupitt** [comment]
... so let's tag this as an enhancement

**Comment by AttilaTheFun** [comment]
@jcupitt thanks for the context - FWIW I just tested the libheif example program heif_enc for which they provide the source code:
https://github.com/strukturag/libheif/blob/master/examples/heif_enc.cc

I was able to produce a good 10-bit HEIF image from a 16-bit PNG (which I produced from a 16-bit CR3 image from my Canon R5).

The size savings are pretty incredible - the CR3 was about 45MB for a 45MP RAW image and the 10-bit HEIF was only about 2.9MB (albeit with the 80% quality setting - I believe lossless would be around 17MB).

I'm not sure how much work it would be to add this functionality to heifsave.c, but it looks like the main change would be changing the heif_image_add_plane call's last argument to take a user-specified bit depth.

**Comment by jcupitt** [comment]
I was able to make a test file with:

```
$ vips colourspace astronauts.png x.png rgb16
$ heif-enc x.png --bit-depth 12 -o x.heic
$ heif-convert x.heic y.png
File contains 1 images
Written to y.png
$ vipsheader y.png
y.png: 900x900 ushort, 3 bands, rgb16, pngload
```

I've added >8 bit load, doing save next. There's a branch here:

https://github.com/libvips/libvips/tree/heif-hdr

**Comment by AttilaTheFun** [comment]
That's awesome @jcupitt! If you manage to get this working, I'd love to add high bit-depth image support to my app with libvips. 
As of the iPhone 12 Pro, iPhones are capable of shooting in ProRaw 12-bit images. (Also most DSLRs / mirrorless cameras  capture raw images with 12-16 bits.)
While the iPhones still have 8 bit displays, the new MacBooks have 10-bit displays as do most higher-end OLED TVs. 
I'd love to allow users to share 10+ bit photos without having to store massive RAW or PNG files.

**Comment by jcupitt** [comment]
OK, all done, I think, I made a PR:

https://github.com/libvips/libvips/pull/2596

Any testing etc. very welcome.

**Comment by jcupitt** [comment]
OK, merged to master, this will be in 8.13. Any testing would be very helpful.
